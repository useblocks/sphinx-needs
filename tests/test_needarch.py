from pathlib import Path

import pytest


@pytest.mark.parametrize("test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needarch"}], indirect=True)
def test_doc_needarch(test_app):
    app = test_app
    app.build()
    html = Path(app.outdir, "index.html").read_text(encoding="utf8")
    assert html


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needarch_negative_tests"}], indirect=True
)
def test_doc_needarch_negative(test_app):
    import subprocess

    app = test_app

    srcdir = Path(app.srcdir)
    out_dir = srcdir / "_build"

    out = subprocess.run(["sphinx-build", "-M", "html", srcdir, out_dir], capture_output=True)

    assert out.returncode == 1
    assert (
        "sphinx_needs.directives.needuml.NeedArchException: Directive needarch "
        "can only be used inside a need." in out.stderr.decode("utf-8")
    )


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needarch_jinja_func_import"}], indirect=True
)
def test_doc_needarch_jinja_import(test_app, snapshot):
    app = test_app
    app.build()
    html = Path(app.outdir, "index.html").read_text(encoding="utf8")
    assert html

    # check needarch
    all_needumls = app.env.needs_all_needumls
    assert all_needumls == snapshot


@pytest.mark.parametrize(
    "test_app", [{"buildername": "html", "srcdir": "doc_test/doc_needarch_jinja_func_need"}], indirect=True
)
def test_needarch_jinja_func_need(test_app, snapshot):
    app = test_app
    app.build()

    all_needumls = app.env.needs_all_needumls
    assert all_needumls == snapshot

    html = Path(app.outdir, "index.html").read_text(encoding="utf8")
    assert "as INT_001 [[../index.html#INT_001]]" in html

    import subprocess

    srcdir = Path(app.srcdir)
    out_dir = srcdir / "_build"

    out = subprocess.run(["sphinx-build", "-M", "html", srcdir, out_dir], capture_output=True)
    assert out.returncode == 0
