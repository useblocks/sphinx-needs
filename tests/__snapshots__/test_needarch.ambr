# serializer version: 1
# name: test_doc_needarch_jinja_import[test_app0]
  dict({
    'needarch-index-0': dict({
      'align': 'center',
      'caption': None,
      'config': '',
      'config_names': None,
      'content': '''
        Alice -> Bob: Hi Bob
        Bob --> Alice: hi Alice
        
        {{import("uses", "tests")}}
      ''',
      'content_calculated': '''
        @startuml
        
        Alice -> Bob: Hi Bob
        Bob --> Alice: hi Alice
        
        node "<size:12>User Story</size>\n**Story**\n<size:10>US_001</size>" as US_001 [[../index.html#US_001]] #BFD8D2
        node "<size:12>User Story</size>\n**Story 002**\n<size:10>US_002</size>" as US_002 [[../index.html#US_002]] #BFD8D2
        card "<size:12>Interface</size>\n**Test interface**\n<size:10>INT_001</size>" as INT_001 [[../index.html#INT_001]] #BFD8D2
        card "<size:12>Interface</size>\n**Test interface2**\n<size:10>INT_002</size>" as INT_002 [[../index.html#INT_002]] #BFD8D2
        
        @enduml
  
      ''',
      'debug': True,
      'docname': 'index',
      'extra': dict({
      }),
      'is_arch': True,
      'key': None,
      'lineno': 3,
      'save': None,
      'scale': '50',
      'target_id': 'needarch-index-0',
    }),
  })
# ---
# name: test_needarch_jinja_func_need[test_app0]
  dict({
    'needarch-index-0': dict({
      'align': 'center',
      'caption': None,
      'config': '',
      'config_names': None,
      'content': '''
        DC -> Marvel: Hi Kevin
        Marvel --> DC: Anyone there?
        
        {{flow(need().id)}}
      ''',
      'content_calculated': '''
        @startuml
        
        DC -> Marvel: Hi Kevin
        Marvel --> DC: Anyone there?
        
        card "<size:12>Interface</size>\n**Test needuml**\n**jinja func need**\n<size:10>INT_001</size>" as INT_001 [[../index.html#INT_001]] #BFD8D2
        @enduml
  
      ''',
      'debug': False,
      'docname': 'index',
      'extra': dict({
      }),
      'is_arch': True,
      'key': None,
      'lineno': 1,
      'save': None,
      'scale': '',
      'target_id': 'needarch-index-0',
    }),
  })
# ---
